### ISSUE TITLE
Exclude files from encryption

### ISSUE THREAD

**Issue by duckpuppy** [TheLocehiliosan/yadm#86]
I have a situation where I want to encrypt a directory because the app I'm using scatters it's configuration across multiple files and subfolders in a single parent folder.  There's a "logs" subfolder, though, and I really don't want that encrypted and committed due to size issues (it's currently 2.6G, and I'm going to clean it out, but it will grow again).

I'm using the current dev branch yadm that has the fix from #81, and it works wonderfully including whole directories otherwise.

**Comment by TheLocehiliosan** [comment]
I think I understand the case you are describing and I have a few ideas about how it could be technically implemented. The user interface would probably be prefacing globs in `.yadm/encrypt` with a `!`.

Until there is a solution implemented, are there few enough directories and files that you can specify them all except the "logs" dir? Something like this to get by:
```
.yadm/encrypt
---------------
app/*.{ext1,ext2,ext3}
app/{dir1,dir2,dir3,dir4}
```

**Comment by duckpuppy** [comment]
I thought I had commented this already, but I also found a workaround that may have some other implications...

Because the file listing is wrapped in an `eval`, you can pipe the output of the `ls` to other commands, so `app/* | grep -v logs` works in this case.  It's useful, but maybe there are security implications?  I can't really think of anything that doesn't involve someone maliciously editing your `.yadm/encrypt` file.

**Comment by TheLocehiliosan** [comment]
Thank you for bringing up this excellent point. I'm not happy using the `eval` of `ls` for glob expansion. I will find an alternative solution.

**Comment by TheLocehiliosan** [comment]
@duckpuppy I just pushed a change to the `dev` branch that will likely break your `|grep -v logs` work-around (or at least it *should*). Just wanted you to have a heads up, as it sounds like you are using the `dev` branch.

I will plan to implement support for prefacing `.yadm/encrypt` lines with `!` to exclude them using the `--exclude` feature of `tar`.  It will take a bit of time to write some test cases and work up the final solution though.

One slight concern I have is, someone *might* have a file that begins with `!` which they want to encrypt (unlikely I know). I wonder if it might be better to have any patterns to be excluded prefaced with `--exclude=`. The likelihood of collision with actual files would drop significantly, and it might be easier to use the whole line as a parameter to `tar`. Do you have any thoughts on the matter? In your case it would look something like this:

```
.yadm/encrypt
---------------
app/*
--exclude=app/logs
```

**Comment by duckpuppy** [comment]
Sorry for the late reply... the `--exclude` format isn't as "pretty" as prefixing with `!`, but I'm not sure that's enough to not do it.  The `.gitignore` file format (to NOT ignore a file, in this case) uses `!` as well, though, so I wonder how often the issue would come up?  If it's not that often, then at the very least the file could be named something else and the bootstrap script could be used to symlink it to it's `!`-prefixed version if it were really necessary.

**Comment by TheLocehiliosan** [comment]
Thanks for the reply. I've actually go back the other way (preferring `!`). I have a mostly working set of tests and changes, I'm close to publishing to the `dev` branch.

**Comment by TheLocehiliosan** [comment]
@duckpuppy This morning I pushed a version to the `dev` branch which adds support for `!<pattern>` exclusions in `.yadm/encrypt`. I'd be interested in how it works for you.

**Comment by duckpuppy** [comment]
I've been terrible at keeping up with GitHub issues lately... sorry.

I'm using the dev version of yadm now, and the exclusions are working perfectly so far.  Thanks for the new feature!

**Comment by TheLocehiliosan** [comment]
This has been released in 1.12.0.
